index processing guidelines (#1189)
google/exposure-notifications-server: docs/getting-started/downloading-export-batches-keys.md

configurable. We recommend batches are configured to be generated at most every
4 hours, with a likely time being somewhere between 4 and 24 hours.

# Consuming Export Batches

## Index File Structure

The index.txt file indicates which export files are available for a mobile device
to consume. The index file contains relative path names for available archives.
A client is typically given two configuration items for pointing at a server

> * The absolute location of the index.txt file

Reviewer: Should we make a note that app developers are encouraged to use runtime configuration and **NOT** hardcode these values?

* The absolute base path to which the relative data in the index file should be appended

As an example. If an index file is located at `https://cdn.example.com/24h/region/index.txt`
and the export base path of `https://cdn.example.com/24h/` and the contents of the index
file are:

```
region/file1.zip
region/file21.zip
region/other_export_.zip
```

Then the following 3 archives are are being offered to clients:

1. `https://cdn.example.com/24h/region/file1.zip`
1. `https://cdn.example.com/24h/region/file21.zip`
1. `https://cdn.example.com/24h/region/other_export_.zip`

And these are being arrived at by appending the contents of the index file (one per line)
to the base path provided by the server operator.

### Index Processing Guideliens

> * Do not make any assumptions or dependencies on the layout of the .zip filenames themselves.

Reviewer: Let's add something about the character set/encoding? US-ASCII, UTF-8

* Do not make any assumptions about the depth of the ralative path in the index file. 
* The index.txt will always be ordered from oldest file to newest file.

> * Each file only needs to be processed once.

Reviewer: \* Do not assume files will end with `.zip`, but _do_ assume that all files are in zip format.

There are two modes of index file processing that meet these guidelines.

#### Last seen method (most common):

The most common method (in use by the Android reference application) is for the client
to remember the last filename that it has seen in an index.txt file.

On next processing, if that line is seen in the file, the application will process

> from the next file to the end of the index. If that lien is not present in the file,

Reviewer: ```suggestion
from the next file to the end of the index. If that line is not present in the file,
```

configurable. We recommend batches are configured to be generated at most every
4 hours, with a likely time being somewhere between 4 and 24 hours.

# Consuming Export Batches

## Index File Structure

The index.txt file indicates which export files are available for a mobile device
to consume. The index file contains relative path names for available archives.
A client is typically given two configuration items for pointing at a server.

> Mobile app developers are strongly encouraged to make these two values remotly

Reviewer 2: ```suggestion
Mobile app developers are strongly encouraged to make these two values remotely
```

configurable so that they can be changed without requiring an application
update.
 
* The absolute location of the index.txt file
* The absolute base path to which the relative data in the index file should be appended

As an example. If an index file is located at `https://cdn.example.com/24h/region/index.txt`
and the export base path of `https://cdn.example.com/24h/` and the contents of the index
file are:

```
region/file1.zip
region/file21.zip
region/other_export_.zip
```

Then the following 3 archives are are being offered to clients:

1. `https://cdn.example.com/24h/region/file1.zip`
1. `https://cdn.example.com/24h/region/file21.zip`
1. `https://cdn.example.com/24h/region/other_export_.zip`

> And these are being arrived at by appending the contents of the index file (one per line)

Reviewer 3: ```suggestion
And these are derived by appending the contents of the index file (one per line)
```
Or constructed? "being arrived at" is a bit hard to read.

to the base path provided by the server operator.

> ### Index Processing Guideliens

Reviewer 2: ```suggestion
### Index Processing Guidelines
```

* Do not make any assumptions about the depth of the relative path in the index file. 
* The HTTP server will return the index.txt file with a content type of `text/plain`
* The index.txt file is encoded using `US-ASCII`
* The index.txt will always be ordered from oldest file to newest file.
* Each file only needs to be processed once.
* Do not assume that archive files end in the `.zip` extension, but they will be valid zip files.

There are two modes of index file processing that meet these guidelines.

#### Last seen method (most common):

The most common method (in use by the Android reference application) is for the client
to remember the last filename that it has seen in an index.txt file.

On next processing, if that line is seen in the file, the application will process
from the next file to the end of the index. If that line is not present in the file,
then the application will restart processing of the index from the beginning (line 0).

When done, the application always remembers the last filename processed.

#### Set method:

The client application should treate absolute URLs of the exports indicated in the the

> index.txt file contents as a set. The client should store the previsuly known state in

Reviewer 2: ```suggestion
index.txt file contents as a set. The client should store the previously known state in
```

local storage.

If differences are seen, the clients should download the items missing from the set in top

> to bottom order (oldest to newest) and hand them to the operating system for matching.

Reviewer 3: ```suggestion
to bottom order (oldest to newest) and provide them to the operating system for matching.
```
or pass, which is used later as the action

The new set of absolute URLs should then be retained for processing next time around.

## Export File Concepts

As the protocol has changed over the deployment of various exposure notifications
installations, we have made a decisions to drop the usage of a batch
being split across multiple files, and show all batches as single file size only.

> This reduces the possilibty that a client app will pass un-parsable batch files

Reviewer 3: ```suggestion
This reduces the possibility that a client app will pass un-parsable batch files
```

to the operating system.

Practically speaking: currently all export files produced by this system will have
`batchNum` of `1` and `batchSize` of `1`. In the event a time period is too 
large to find in one file according to the max records setting, there will
be multiple files with the same `startTimestamp` but different `endTimestamps`. This
is done to ensure successful processing based on how both iOS and Android
handle export processing.

> # Configurating Export Batches

Reviewer 2: ```suggestion
# Configuring Export Batches
```
